--- a/microscript/ILibDuktape_HttpStream.c
+++ b/microscript/ILibDuktape_HttpStream.c
@@ -3222,8 +3222,6 @@
 
 void ILibDuktape_HttpStream_IMSG_EndSink2(duk_context *ctx, void ** args, int argsLen)
 {
-	void *tmp = args[1];
-	if (ILibMemory_CanaryOK(tmp))
 	{
 		duk_push_heapptr(ctx, args[0]);							// [IMSG]
 		duk_prepare_method_call(ctx, -1, "removeAllListeners");	// [IMSG][removeAll][this]
@@ -3234,9 +3232,11 @@
 duk_ret_t ILibDuktape_HttpStream_IMSG_EndSink(duk_context *ctx)
 {
 	duk_push_this(ctx);
-	void *t = Duktape_PushBuffer(ctx, 1); duk_put_prop_string(ctx, -2, "tmp");
 	void *h = duk_get_heapptr(ctx, -1);
-	ILibDuktape_Immediate(ctx, (void*[]) { h, t }, 2, ILibDuktape_HttpStream_IMSG_EndSink2);
+	void *imm = ILibDuktape_Immediate(ctx, (void*[]) { h }, 1, ILibDuktape_HttpStream_IMSG_EndSink2);
+	duk_push_heapptr(ctx, imm);
+	duk_dup(ctx, -2);
+	duk_put_prop_string(ctx, -2, "\xFF_incomingMessage");
 	return(0);
 }
 duk_ret_t ILibDuktape_HttpStream_md5(duk_context *ctx)
@@ -3256,6 +3256,7 @@
 	duk_context *ctx = data->DS->writableStream->ctx;
 	uintptr_t ctxnonce = duk_ctx_nonce(ctx);
 	int top = duk_get_top(ctx);
+	int isClientResponse = header != NULL && header->Directive == NULL;
 
 	if (data->bodyStream != NULL)
 	{
@@ -3540,7 +3541,7 @@
 			if (duk_pcall_method(ctx, 1) != 0) ILibDuktape_Process_UncaughtExceptionEx(ctx, "httpStream.onReceive(): Error dispatching 'end': %s", duk_safe_to_string(ctx, -1));
 			duk_pop(ctx);													// ...
 
-			if (header->Directive == NULL && data->connectionCloseSpecified == 0)
+			if (isClientResponse && data->connectionCloseSpecified == 0)
 			{
 				duk_push_heapptr(ctx, data->DS->ParentObject);						// [httpStream]
 				if (duk_has_prop_string(ctx, -1, ILibDuktape_HTTPStream2Socket))
