--- a/microscript/ILibDuktape_Polyfills.c
+++ b/microscript/ILibDuktape_Polyfills.c
@@ -1183,12 +1183,9 @@
 	}
 	else
 	{
-		if (ptrs->timerType == ILibDuktape_Timer_Type_IMMEDIATE)
-		{
-			duk_push_heap_stash(ctx);
-			duk_del_prop_string(ctx, -1, Duktape_GetStashKey(ptrs->object));
-			duk_pop(ctx);
-		}
+		duk_push_heap_stash(ctx);
+		duk_del_prop_string(ctx, -1, Duktape_GetStashKey(ptrs->object));
+		duk_pop(ctx);
 
 		duk_del_prop_string(ctx, -2, "\xFF_callback");
 		duk_del_prop_string(ctx, -2, "\xFF_argArray");
@@ -1236,11 +1233,6 @@
 	case ILibDuktape_Timer_Type_IMMEDIATE:
 		ILibDuktape_WriteID(ctx, "Timers.immediate");	
 		metadata = "setImmediate()";
-		// We're only saving a reference for immediates
-		duk_push_heap_stash(ctx);															//[retVal][stash]
-		duk_dup(ctx, -2);																	//[retVal][stash][immediate]
-		duk_put_prop_string(ctx, -2, Duktape_GetStashKey(duk_get_heapptr(ctx, -1)));		//[retVal][stash]
-		duk_pop(ctx);																		//[retVal]
 		break;
 	case ILibDuktape_Timer_Type_INTERVAL:
 		ILibDuktape_WriteID(ctx, "Timers.interval");
@@ -1279,6 +1271,10 @@
 		ILibLifeTime_AddEx3(ILibGetBaseTimer(chain), ptrs, ptrs->timeout, ILibDuktape_Polyfills_timer_elapsed, NULL, metadata));
 	duk_put_prop_string(ctx, -2, "\xFF_token");
 	ILibDuktape_CreateEventWithSetterEx(ctx, "metadata", ILibDuktape_Polyfills_Timer_Metadata);
+	duk_push_heap_stash(ctx);															//[retVal][stash]
+	duk_dup(ctx, -2);																	//[retVal][stash][timer]
+	duk_put_prop_string(ctx, -2, Duktape_GetStashKey(duk_get_heapptr(ctx, -1)));		//[retVal][stash]
+	duk_pop(ctx);																		//[retVal]
 	return 1;
 }
 duk_ret_t ILibDuktape_Polyfills_timer_clear(duk_context *ctx)
@@ -1309,14 +1305,11 @@
 	duk_get_prop_string(ctx, 0, ILibDuktape_Timer_Ptrs);
 	ptrs = (ILibDuktape_Timer*)Duktape_GetBuffer(ctx, -1, NULL);
 
-	if (ptrs->timerType == ILibDuktape_Timer_Type_IMMEDIATE)
-	{
-		duk_push_heap_stash(ctx);
-		duk_del_prop_string(ctx, -1, Duktape_GetStashKey(ptrs->object));
-		duk_pop(ctx);
-	}
-
+	void *timerObject = ptrs->object;
 	ILibLifeTime_Remove(ILibGetBaseTimer(Duktape_GetChain(ctx)), ptrs);
+	duk_push_heap_stash(ctx);
+	duk_del_prop_string(ctx, -1, Duktape_GetStashKey(timerObject));
+	duk_pop(ctx);
 	return 0;
 }
 void ILibDuktape_Polyfills_timer(duk_context *ctx)
